File: app/src/main/java/com/hover/stax/hover/HoverSession.java
package com.hover.stax.hover;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

import androidx.fragment.app.Fragment;

import com.hover.sdk.actions.HoverAction;
import com.hover.sdk.api.Hover;
import com.hover.sdk.api.HoverParameters;
import com.hover.stax.R;
import com.hover.stax.channels.Channel;
import com.hover.stax.contacts.PhoneHelper;
import com.hover.stax.settings.KeyStoreExecutor;
import com.hover.stax.utils.Constants;
import com.hover.stax.utils.Utils;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;
import java.util.List;

final public class HoverSession {
    private final static String TAG = "HoverSession";

    final private Fragment frag;
    final private Channel channel;
    final private int requestCode, finalScreenTime;


    private HoverSession(Builder b) {
        Hover.setPermissionActivity(Constants.PERM_ACTIVITY, b.activity);
        frag = b.fragment;
        channel = b.channel;
        requestCode = b.requestCode;
        finalScreenTime = b.finalScreenTime;
        HoverParameters.Builder builder = getBasicBuilder(b);
        addExtras(builder, b.extras, b.action);
//		addPin(builder, b.activity); // Not active
        startHover(builder, b.activity);
    }

    private HoverParameters.Builder getBasicBuilder(Builder b) {
        HoverParameters.Builder builder = new HoverParameters.Builder(b.activity);
        builder.setEnvironment(Utils.getBoolean(Constants.TEST_MODE, b.activity) ? HoverParameters.TEST_ENV : HoverParameters.PROD_ENV);
        builder.request(b.action.public_id);
        builder.setHeader(getMessage(b.action, b.activity));
        builder.initialProcessingMessage("");
        builder.showUserStepDescriptions(true);
        builder.finalMsgDisplayTime(finalScreenTime);
        builder.style(R.style.StaxHoverTheme);
        builder.sessionOverlayLayout(R.layout.stax_transacting_in_progress);
        return builder;
    }

    private void addExtras(HoverParameters.Builder builder, JSONObject extras, HoverAction action) {
        List<String> required_extras = action.getRequiredParams();
        Iterator<?> keys = extras.keys();
        while (keys.hasNext()) {
            String key = (String) keys.next();
            String normalizedVal = parseExtra(key, extras.optString(key), required_extras);
            if (normalizedVal != null)
                builder.extra(key, normalizedVal);
        }
    }

    private String parseExtra(String key, String value, List<String> requiredExtras) {
        if (value == null || !requiredExtras.contains(key)) {
            return null;
        }
        if (key.equals(HoverAction.PHONE_KEY)) {
            return PhoneHelper.normalizeNumberByCountry(value, channel.countryAlpha2);
        }
        return value;
    }

    private void addPin(HoverParameters.Builder builder, Activity a) {
        builder.extra(HoverAction.PIN_KEY, KeyStoreExecutor.decrypt(channel.pin, a));
    }

    private String getMessage(HoverAction a, Context c) {
        switch (a.transaction_type) {
            case HoverAction.BALANCE:
                return c.getString(R.string.balance_msg, a.from_institution_name);
            case HoverAction.AIRTIME:
                return c.getString(R.string.airtime_msg);
            default:
                return c.getString(R.string.transfer_msg);
        }
    }

    private void startHover(HoverParameters.Builder builder, Activity a) {
        Intent i = builder.buildIntent();
        Utils.logAnalyticsEvent(a.getString(R.string.start_load_screen), a);
        if (frag != null)
            frag.startActivityForResult(i, requestCode);
        else
            a.startActivityForResult(i, requestCode);
    }

    public static class Builder {
        private final Activity activity;
        private Fragment fragment;
        private Channel channel;
        private HoverAction action;
        private JSONObject extras;
        private int requestCode, finalScreenTime = 4000;

        public Builder(HoverAction a, Channel c, Activity activity, int code) {
            if (a == null) throw new IllegalArgumentException("Action must not be null");
            this.activity = activity;
            channel = c;
            action = a;
            extras = new JSONObject();
            requestCode = code;
        }

        public Builder(HoverAction a, Channel c, Activity act, int requestCode, Fragment frag) {
            this(a, c, act, requestCode);
            fragment = frag;
        }

        public HoverSession.Builder extra(String key, String value) {
            try {
                extras.put(key, value);
            } catch (JSONException e) {
                Log.e(TAG, "Failed to add extra");
            }
            return this;
        }

        public HoverSession.Builder finalScreenTime(int ms) {
            finalScreenTime = ms;
            return this;
        }

        public HoverSession run() {
            return new HoverSession(this);
        }
    }
}

File: app/src/main/java/com/hover/stax/hover/HoverSession.kt
/*
 * Copyright 2022 Stax
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.hover.stax.hover

import android.app.Activity
import android.content.Intent
import androidx.fragment.app.Fragment
import com.hover.sdk.actions.HoverAction
import com.hover.sdk.api.Hover
import com.hover.sdk.api.HoverParameters
import com.hover.stax.R
import com.hover.stax.contacts.PhoneHelper
import com.hover.stax.domain.model.ACCOUNT_ID
import com.hover.stax.domain.model.ACCOUNT_NAME
import com.hover.stax.domain.model.Account
import com.hover.stax.settings.TEST_MODE
import com.hover.stax.utils.Utils
import org.json.JSONException
import org.json.JSONObject
import timber.log.Timber

const val PERM_ACTIVITY = "com.hover.stax.permissions.PermissionsActivity"
private const val TIMER_LENGTH = 35000

class HoverSession private constructor(b: Builder) {
    private val frag: Fragment?
    private val account: Account
    private val action: HoverAction
    private val finalScreenTime: Int

    private val builder: HoverParameters.Builder

    private fun getBasicBuilder(b: Builder): HoverParameters.Builder = HoverParameters.Builder(b.activity)
        .apply {
            setEnvironment(if (Utils.getBoolean(TEST_MODE, b.activity)) HoverParameters.TEST_ENV else HoverParameters.PROD_ENV)
            extra(ACCOUNT_NAME, account.getAccountNameExtra())
            private_extra(ACCOUNT_ID, account.id.toString())
            request(b.action.public_id)
            setHeader(getMessage(b))
            initialProcessingMessage("")
            showUserStepDescriptions(true)
            timeout(TIMER_LENGTH)
            finalMsgDisplayTime(finalScreenTime)
            style(R.style.StaxHoverTheme)
            sessionOverlayLayout(R.layout.stax_transacting_in_progress)
        }

    private fun addExtras(builder: HoverParameters.Builder, extras: JSONObject) {
        val keys: Iterator<*> = extras.keys()
        while (keys.hasNext()) {
            val key = keys.next() as String
            val normalizedVal = parseExtra(key, extras.optString(key))
            if (normalizedVal != null) builder.extra(key, normalizedVal)
        }
    }

    private fun parseExtra(key: String, value: String?): String? {
        if (value == null) return null
        return if (key == HoverAction.PHONE_KEY) {
            PhoneHelper.normalizeNumberByCountry(value, action.country_alpha2, action.to_country_alpha2)
        } else value
    }

    private fun getMessage(b: Builder): String {
        return if (b.message != null) { b.message!! }
            else {
                when (b.action.transaction_type) {
                    HoverAction.BALANCE -> b.activity.getString(R.string.balance_msg, b.action.from_institution_name)
                    HoverAction.AIRTIME -> b.activity.getString(R.string.airtime_msg)
                    else -> b.activity.getString(R.string.transfer_msg)
                }
            }
    }

    private fun stopEarly(builder: HoverParameters.Builder, varName: String?) {
        if (varName != null && action.output_params.has(varName))
            builder.stopAt(action.output_params.getInt(varName))
    }

    private fun getIntent(): Intent {
        return builder.buildIntent()
    }

    class Builder(a: HoverAction?, c: Account, activity: Activity) {
        val activity: Activity
        var fragment: Fragment? = null
        val account: Account
        var message: String? = null
        val action: HoverAction
        val extras: JSONObject
        var finalScreenTime = 0
        var stopVar: String? = null

        constructor(action: HoverAction,
                    c: Account,
                    extras: HashMap<String, String>?,
                    act: Activity) : this(action, c, act) {
            if (!extras.isNullOrEmpty()) { extras(extras) }
        }

        constructor(a: HoverAction?, c: Account, act: Activity, frag: Fragment?) : this(a, c, act) {
            fragment = frag
        }

        fun extra(key: String, value: String?): Builder {
            try {
                extras.put(key, value)
            } catch (e: JSONException) {
                Timber.e("Failed to add extra")
            }
            return this
        }

        fun extras(es: HashMap<String, String>): Builder {
            try {
                es.forEach { extras.put(it.key, it.value) }
            } catch (e: JSONException) {
                Timber.e("Failed to add extra")
            }
            return this
        }

        fun stopAt(varName: String): Builder {
            stopVar = varName
            return this
        }

        fun finalScreenTime(ms: Int): Builder {
            finalScreenTime = ms
            return this
        }

        fun message(msg: String): Builder {
            message = msg
            return this
        }

        fun build(): Intent {
            return HoverSession(this).getIntent()
        }

        init {
            requireNotNull(a) { "Action must not be null" }
            this.activity = activity
            account = c
            action = a
            extras = JSONObject()
        }
    }

    init {
        Hover.setPermissionActivity(PERM_ACTIVITY, b.activity)
        frag = b.fragment
        account = b.account
        action = b.action
        finalScreenTime = b.finalScreenTime

        builder = getBasicBuilder(b)
        addExtras(builder, b.extras)
        stopEarly(builder, b.stopVar)
    }
}